### ISSUE TITLE
Consider 1-based indexing for tensor [ operator

### ISSUE THREAD

**Issue by jjallaire** [rstudio/tensorflow#195]
I've recently been working on some Keras examples that mix use of R/NumPy arrays and raw TF tensor objects. Within the code the same data is accessed via the `[` operator in both contexts. However, one context is 0-based and the other is 1-based. 

This makes for an incredibly confusing situation, and that's for me the maintainer of all of these packages! I wonder if we should revisit our decision to use 0-based indexing for the  tensor`[` operator. The original motivation was two-fold:

1) The other lower-level `tf$` functions use 0-based indexing; and

2) If you were porting Python code that used 0-based indexing then this would ease the porting.

However, I wonder if the case of mixed R array and TF tensor access will be more common. I think it's a reasonably clean story to say that *as a convenience* we make TF tensors 1-based when using the R `[` operator. This is in fact more semantically consistent with `[`, in which AFAICT 0 is equivalent to dropping a dimension.

If we did this we could detect 0 based indexes and yield an error from them (with a note describing how to put the package in backward compatibility mode via an option). I realize that this would cause some short-term pain, but if 1-based is really the right long term direction it's much better that we do it now.

@goldingn I am particularly interested in your opinion about this.

**Comment by jjallaire** [comment]
Here's the original discussion for reference: https://github.com/rstudio/tensorflow/pull/19

Note that we are already inconsistent with Python by being end-inclusive. It seems like being fully consistent with R `[` would be more clear. No matter what we do some people will be confused some of the time, but to me I can't think of anything as bad as writing code that mixes 0 and 1 based indexing!

@dfalbel Would be interesting to get your take here as well.

**Comment by goldingn** [comment]
Yeah, I suspect being fully consistent with R will increase uptake among R users. I don't think we'll be able to detect and warn on all 0-based indices, but even if it's a breaking change, I think it'll be worth it in the long run.

There are a lot of weird quirks to R's extract and replace syntax (e.g. using logicals; extracting by cell number; extracting by a matrix with `dim` columns), and people seem to use it in very different ways. So replicating all of that manually would be a ton of work.

In [greta](https://greta-dev.github.io/greta/) I've implemented an exact copy of R's syntax by executing the extract call on a dummy array, with elements giving the cell-numbers as indexed in python. That yields the information needed to [extract and reshape tensors](https://github.com/greta-dev/greta/blob/master/R/tf_functions.R#L222-L231) using `tf$reshape()` and `tf$gather`. Though I haven't looked into the performance implications of this rather than the `tf$strided_slice` we currently use.

If we wanted to go down that route, I'd be happy to move the code over to tensorflow; then I could drop it from greta.

**Comment by goldingn** [comment]
Of course, users will still have to know about 0-indexing when passing arguments directly to TF functions (e.g. `tf$reduce_sum(x, axis = 0)`), so this won't be a panacea.

**Comment by jjallaire** [comment]
Yes, I originally thought that this inconsistency (functions args vs. `[`) would be confusing but I think that's the less of the two evils.

It would be *great* if you could move the code from Greta into the TF package. The one other thing I think we would need would be some mechanism to opt back into 0-based indexing (probably an R option) and some way to proactively warn users that their code is "broken" (and letting them know about the option to revert to 0-based indexing). I realize we wouldn't be able to catch all cases, but would we catch at least some?

**Comment by jjallaire** [comment]
With respect to how we stage this, I'm actually working on a book draft that will want to use the revised indexing scheme (attempting to explain the mixed-indexing in the book was actually the motivation for me raising this!). If moving the work over from Greta is a bit more involved I think we should patch what we have and then do that later so that I can in turn make the revisions for the book draft (which is due to the publisher very soon).

**Comment by bwlewis** [comment]
I think an option to set  base index is a good idea. Plus it's very
nostalgic like Fortran!

**Comment by dfalbel** [comment]
I also think that it's a great change! It will be much more intuitive to use. 
I spent a lot of time implementing this function correctly! :)

# the 3rd loss function, total variation loss,
# designed to keep the generated image locally coherent

```
total_variation_loss <- function(x){
  y_ij  <- x[,0:(img_nrows - 2L), 0:(img_ncols - 2L),]
  y_i1j <- x[,1:(img_nrows - 1L), 0:(img_ncols - 2L),]
  y_ij1 <- x[,0:(img_nrows - 2L), 1:(img_ncols - 1L),]
  
  a <- K$square(y_ij - y_i1j)
  b <- K$square(y_ij - y_ij1)
  K$sum(K$pow(a + b, 1.25))
}
```

In this keras example: https://keras.rstudio.com/articles/examples/neural_style_transfer.html

**Comment by jjallaire** [comment]
@dfalbel Is that the only Keras example that uses the tensor extract operator? (want to make sure that we modify all of them that we need to after the change goes in).

**Comment by goldingn** [comment]
@jjallaire implementing the greta behaviour looked easier than trying to understand how to allow 1-based indexing in my 0-based indexing code, so I've implemented that over on #199. Seems to work fine, so hopefully that'll be good for meeting your deadline?

The only issue is that passing indices with 0s is perfectly valid for R's extraction syntax, so I don't think it'll be possible to warn users that they're using the old API without breaking the new syntax.

**Comment by goldingn** [comment]
Relatedly, but I'm guessing less urgently, it would be fairly straightforward to implement R's complete replace syntax for tensors. I.e.:

```r
x <- tf$constant(1:3)
x[2:3] <- 0
```

That would just be a straight port from greta too. I probably won't have time to do that for a couple of weeks, but feel free to open a new issue on that topic if you'd be keen for that functionality.

**Comment by jjallaire** [comment]
@dfalbel I have merged the PR that migrates to 1-based indexing. Would you update the Keras example you referenced (and any other examples that rely on the old behavior)?

**Comment by jjallaire** [comment]
I see from the implementation now that it's not really practical to detect when 0 based indexing is being used (or is there a way I'm not seeing?).

**Comment by jjallaire** [comment]
@jbaker92 FYI in case sgmcmc relies on 0-based indexing anywhere (it doesn't look like you to as I ran your tests under both indexing schemes and in both cases 4 tests fails, presumably these failures are unrelated to the indexing)

**Comment by jjallaire** [comment]
Docs updated here: https://tensorflow.rstudio.com/tensorflow/articles/using_tensorflow_api.html#tensor-extraction

**Comment by jbaker92** [comment]
Thanks @jjallaire pretty sure there's no 0-based indexing but will double check.
